internal/wasm/module_context_test.go

```diff
@@ -0,0 +1,199 @@
+		},
+		{
+			name:       "ctx->nil: same",
+			mod:        &ModuleContext{ctx: context.Background()},
+			ctx:        nil,
+			expectSame: true,
+		},
+		{
+			name:       "ctx1->ctx2: not same",
+			mod:        &ModuleContext{ctx: context.Background()},
+			ctx:        context.WithValue(context.Background(), key("a"), "b"),
+			expectSame: false,
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			mod2 := tc.mod.WithContext(tc.ctx)
+			if tc.expectSame {
+				require.Same(t, tc.mod, mod2)
+			} else {
+				require.NotSame(t, tc.mod, mod2)
+				require.Equal(t, tc.ctx, mod2.Context())
+			}
+		})
+	}
+}
+
+func TestModuleContext_WithMemory(t *testing.T) {
+	tests := []struct {
+		name       string
+		mod        *ModuleContext
+		mem        *MemoryInstance
+		expectSame bool
+	}{
+		{
+			name:       "nil->nil: same",
+			mod:        &ModuleContext{},
+			mem:        nil,
+			expectSame: true,
+		},
+		{
+			name:       "nil->mem: not same",
+			mod:        &ModuleContext{},
+			mem:        &MemoryInstance{},
+			expectSame: false,
+		},
+		{
+			name:       "mem->nil: same",
+			mod:        &ModuleContext{memory: &MemoryInstance{}},
+			mem:        nil,
+			expectSame: true,
+		},
+		{
+			name:       "mem1->mem2: not same",
+			mod:        &ModuleContext{memory: &MemoryInstance{}},
+			mem:        &MemoryInstance{},
+			expectSame: false,
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			mod2 := tc.mod.WithMemory(tc.mem)
+			if tc.expectSame {
+				require.Same(t, tc.mod, mod2)
+			} else {
+				require.NotSame(t, tc.mod, mod2)
+				require.Equal(t, tc.mem, mod2.memory)
+			}
+		})
+	}
+}
+
+func TestModuleContext_String(t *testing.T) {
+	s := newStore()
+
+	tests := []struct {
+		name, moduleName, expected string
+	}{
+		{
+			name:       "empty",
+			moduleName: "",
+			expected:   "Module[]",
+		},
+		{
+			name:       "not empty",
+			moduleName: "math",
+			expected:   "Module[math]",
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			// Ensure paths that can create the host module can see the name.
+			m, err := s.Instantiate(context.Background(), &Module{}, tc.moduleName, nil)
+			defer m.Close() //nolint
+
+			require.NoError(t, err)
+			require.Equal(t, tc.expected, m.String())
+			require.Equal(t, tc.expected, s.Module(m.module.Name).String())
+		})
+	}
+}
+
+func TestModuleContext_Close(t *testing.T) {
+	s := newStore()
+
+	t.Run("calls store.CloseModule(module.name)", func(t *testing.T) {
+		moduleName := t.Name()
+		m, err := s.Instantiate(context.Background(), &Module{}, moduleName, nil)
+		require.NoError(t, err)
+
+		// We use side effects to determine if Close in fact called store.CloseModule (without repeating store_test.go).
+		// One side effect of store.CloseModule is that the moduleName can no longer be looked up. Verify our base case.
+		require.Equal(t, s.Module(moduleName), m)
```

Comment:
@nullpo-head here's an example of using other tests to build higher level signals that something happened.

For example, through this and other tests, we know that `Module.Close()` ends up resulting in `runtime.Module(moduleName) == nil` (provided you check the base case). This can lead to simpler tests when we are trying to see if a function that calls close executed. e.g. try to lookup the module it was called from.